### PR TITLE
core,services: add ChannelzService class

### PR DESCRIPTION
**Pull request by zpencer**
This implements the class to expose the stats as
protos. GetServerSockets is still unimplemented and will require a
follow up change.

This PR is large but more than half of the LOC are tests.

### CODE REVIEW

File: services/src/main/java/io/grpc/services/ChannelzProtoUtil.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    Address.Builder builder = Address.newBuilder();
+    if (address instanceof InetSocketAddress) {
+      InetSocketAddress inetAddress = (InetSocketAddress) address;
+      builder.setTcpipAddress(
+          TcpIpAddress
+              .newBuilder()
+              .setIpAddress(
+                  ByteString.copyFrom(inetAddress.getAddress().getAddress()))
+              .build());
+    } else if (address.getClass().getName().equals("io.netty.channel.unix.DomainSocketAddress")) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzProtoUtil.java]: This should be an `endsWith` to account for shading.

File: services/src/main/java/io/grpc/services/ChannelzProtoUtil.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        .setState(toState(stats.state))
+        .setCallsStarted(stats.callsStarted)
+        .setCallsSucceeded(stats.callsSucceeded)
+        .setCallsFailed(stats.callsFailed)
+        .setLastCallStartedTimestamp(Timestamps.fromMillis(stats.lastCallStartedMillis))
+        .build();
+  }
+
+  static State toState(ConnectivityState state) {
+    if (state == ConnectivityState.CONNECTING) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzProtoUtil.java]: Enum.valueOf(State.class, state.name());

File: services/src/main/java/io/grpc/services/ChannelzService.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+import io.grpc.channelz.v1.GetTopChannelsResponse;
+import io.grpc.internal.Channelz;
+import io.grpc.internal.Channelz.ChannelStats;
+import io.grpc.internal.Channelz.ServerList;
+import io.grpc.internal.Channelz.ServerStats;
+import io.grpc.internal.Channelz.SocketStats;
+import io.grpc.internal.Instrumented;
+import io.grpc.stub.StreamObserver;
+
+public class ChannelzService extends ChannelzGrpc.ChannelzImplBase {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzService.java]: final

File: services/src/main/java/io/grpc/services/ChannelzService.java [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  private static final int DEFAULT_MAX_PAGE_SIZE = 25;
+  private final Channelz channelz;
+  private final int maxPageSize;
+
+  public ChannelzService() {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzService.java]: since this is public api, lets make this a factory.

File: services/src/main/java/io/grpc/services/ChannelzService.java [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+          = channelz.getRootChannels(request.getStartChannelId(), maxPageSize);
+      GetTopChannelsResponse.Builder responseBuilder = GetTopChannelsResponse
+          .newBuilder()
+          .setEnd(rootChannels.end);
+      for (Instrumented<ChannelStats> c : rootChannels.channels) {
+        responseBuilder.addChannel(ChannelzProtoUtil.toChannel(c));
+      }
+      responseObserver.onNext(responseBuilder.build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzService.java]: let this propagate

File: services/src/main/java/io/grpc/services/ChannelzService.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        return;
+      }
+
+      responseObserver.onNext(
+          GetChannelResponse
+              .newBuilder()
+              .setChannel(ChannelzProtoUtil.toChannel(s))
+              .build());
+      responseObserver.onCompleted();
+    } catch (Throwable t) {
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzService.java]: same here.  It would be better to fail hard, rather than silently.

File: services/src/main/java/io/grpc/services/ChannelzService.java [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+      responseObserver.onError(t);
+    }
+  }
+
+  @Override
+  public void getServerSockets(
+      GetServerSocketsRequest request, StreamObserver<GetServerSocketsResponse> responseObserver) {
+    // TODO(zpencer): fill this one out after refactoring channelz class
+    throw new UnsupportedOperationException();
```
Comment (carl-mastrangelo) [inline services/src/main/java/io/grpc/services/ChannelzService.java]: Status.UNIMPLEMENTED.asRuntimeException() ?

File: services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+  }
+
+  @Test
+  public void toSocketRef() {
+    assertEquals(socketRef, ChannelzProtoUtil.toSocketRef(socket));
+  }
+
+  @Test
+  public void toState() {
+    assertEquals(State.CONNECTING, ChannelzProtoUtil.toState(ConnectivityState.CONNECTING));
```
Comment (carl-mastrangelo) [inline services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java]: this should be a for loop

File: services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+  @Test
+  public void toSocketData() {
+    assertEquals(
+        socketData,
+        ChannelzProtoUtil.toSocketData(socket.transportStats));
+  }
+
+  @Test
+  public void toAddress_inet() throws Exception {
+    InetSocketAddress inet4 = new InetSocketAddress(Inet4Address.getByName("10.0.0.1"), 1000);
```
Comment (carl-mastrangelo) [inline services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java]: does this work for inprocess?

File: services/src/main/java/io/grpc/services/ChannelzProtoUtil.java (hunk 1 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzProtoUtil.java]: done

File: services/src/main/java/io/grpc/services/ChannelzProtoUtil.java (hunk 2 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzProtoUtil.java]: done

File: services/src/main/java/io/grpc/services/ChannelzService.java (hunk 3 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzService.java]: done

File: services/src/main/java/io/grpc/services/ChannelzService.java (hunk 4 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzService.java]: good point about public api, I've also added `@ExperimentalApi` to the class

File: services/src/main/java/io/grpc/services/ChannelzService.java (hunk 5 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzService.java]: Since we are unwrapping `ListenableFuture`, I will need to catch the exceptions somewhere. I moved it into `ChannelzProtoUtil` to just throw a StatusRuntimeException if any exceptions are thrown.

File: services/src/main/java/io/grpc/services/ChannelzService.java (hunk 6 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzService.java]: done

File: services/src/main/java/io/grpc/services/ChannelzService.java (hunk 7 above)
Comment (zpencer) [reply services/src/main/java/io/grpc/services/ChannelzService.java]: done

File: services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java (hunk 8 above)
Comment (zpencer) [reply services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java]: done

File: services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java (hunk 9 above)
Comment (zpencer) [reply services/src/test/java/io/grpc/services/ChannelzProtoUtilTest.java]: `InProcessTransport` does not keep stats right now, but it will eventually. Specifically for sockets, `ChannelzProtoUtil.toAddress` will turn non UDS, non inet sockets into unknown sockets. So we just need to make sure `InProcessSocketAddress` has a properly working `toString`.

**Comment by zpencer** [comment]
Moved some code out of `ChannelzService` so that all heavy lifting is done by either `ChannelzProtoUtil` or `Channelz` itself. This is needed to test `GetServerSockets` more easily, because that service method has some interesting edge cases that are easier to unit test at the low level.
